File: python/sfo_54_the_k-th_largest_node_of_a_binary_search_tree_s1.py
'''
File: sfo_54_the_k-th_largest_node_of_a_binary_search_tree_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def kthLargest(self, root: TreeNode, k: int) -> int:
        def dfs(root):
            if not root: return
            dfs(root.right)
            if self.k == 0: return
            self.k -= 1
            if self.k == 0: self.res = root.val
            dfs(root.left)

        self.k = k
        dfs(root)
        return self.res

# ======= Test Case =======
root = list_to_tree([3, 1, 4, None, 2, None, None, None, None])
k = 1
# ====== Driver Code ======
slt = Solution()
res = slt.kthLargest(root, k)
print(res)

File: cpp/sfo_54_the_kth_largest_node_of_a_binary_search_tree_s1/sfo_54_the_kth_largest_node_of_a_binary_search_tree_s1.cpp
/*
* File: sfo_54_the_kth_largest_node_of_a_binary_search_tree_s1.cpp
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    int kthLargest(TreeNode* root, int k) {
        this->k = k;
        dfs(root);
        return res;
    }
private:
    int res, k;
    void dfs(TreeNode* root) {
        if(root == nullptr) return;
        dfs(root->right);
        if(k == 0) return;
        if(--k == 0) res = root->val;
        dfs(root->left);
    }
};

int main() {
    // ======= Test Case =======
    TreeNode* root = vectorToTree(vector<int> { 3, 1, 4, INT_MAX, 2, INT_MAX, INT_MAX, INT_MAX, INT_MAX });
    int k = 1;
    // ====== Driver Code ======
    Solution* slt = new Solution();
    int res = slt->kthLargest(root, k);
    cout << res << endl;
    
    return 0;
}
